### PR TITLE
Artist's draw method prevents rasterization by default

### PR DESCRIPTION
**Pull request by leejjoon**
## PR Summary

This is another PR to address #23999 
(An alternative PR was proposed #24473 but was not favored as it prevents merging rasterized artists together)

It utilizes `Artist.__init_subclass__`  to decorate its `draw` method by defaullt with `_prevent_rasterization`, so that the rasterization can be properly stopped when an artist that does not support rasterization is encountered. 

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [N/A] New plotting related features are documented with examples.

**Release Notes**
- [N/A] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [N/A] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

File: lib/matplotlib/artist.py [hunk 1]
```diff
@@ -119,6 +143,8 @@ def __init_subclass__(cls):
         cls.set.__qualname__ = f"{cls.__qualname__}.set"
         cls._update_set_signature_and_docstring()
 
+        cls.draw = _prevent_rasterization(cls.draw)
```
Comment (tacaswell) [inline lib/matplotlib/artist.py:146]: Put this above L133 to make sure this still gets run even if the conditional wants to short-circuit us.

File: lib/matplotlib/tests/test_backend_svg.py [hunk 2]
```diff
@@ -122,6 +122,27 @@ def test_rasterized_ordering(fig_test, fig_ref):
     ax_test.plot(x+1, y, "-", c="b", lw=10, rasterized=False, zorder=1.2)
 
 
+@check_figures_equal(tol=5)
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_backend_svg.py:125]:
```suggestion
@check_figures_equal(tol=5, extensions=['svg', 'pdf'])
```

Do not check this with Agg as I do not think it matters in that case.

File: lib/matplotlib/artist.py [hunk 3]
```diff
@@ -20,6 +20,30 @@
...
+def _prevent_rasterization(draw):
+    # we assume that by default artists are not allowed to rasterize (unless
+    # its draw method is explicitly decorated). If it is being drawn after a
+    # rasterized artist and it has reached the rater_depth of 0. We stop
+    # rasterization so that it does not affect the behavior of normal artist
+    # (e.g., change in dpi). If the artist's draw method is decorated
+    # (draw._supports_rasterization is True), it won't  be decorated by
+    # `_prevent_rasterization`.
+
+    if hasattr(draw, "_supports_rasterization"):
```
Comment (tacaswell) [inline lib/matplotlib/artist.py:32]: Do we also want to set `draw._supports_rasterization = False` in this wrapper?

Comment (leejjoon) [reply lib/matplotlib/artist.py:32]: I have no preference, as far as we use the current condition of `getattr(self.draw, "_supports_rasterization", False)` .
And not sure which is better.

**Comment by tacaswell** [comment]
👍🏻 I think this is on the right path.

File: lib/matplotlib/artist.py (hunk 3 above)
Comment (tacaswell) [reply lib/matplotlib/artist.py:33]: On one hand, not setting it is the safer minimal change, but setting it to False (and finding + fixing our internal references to use `getattr(obj, '_supports_rasterization', False)` instead) makes everything more consistent and understandable for the next set of people.

Comment (tacaswell) [reply lib/matplotlib/artist.py:33]: I will not block merging this over sorting that out.

**Review by tacaswell — APPROVED** [review]

File: lib/matplotlib/artist.py [hunk 4]
```diff
@@ -106,6 +131,8 @@ def __init_subclass__(cls):
         # Inject custom set() methods into the subclass with signature and
         # docstring based on the subclasses' properties.
 
+        cls.draw = _prevent_rasterization(cls.draw)
```
Comment (QuLogic) [inline lib/matplotlib/artist.py:134]: Either the above comment needs updating, or this should not go between it and the code below.

File: lib/matplotlib/artist.py [hunk 5]
```diff
@@ -21,6 +21,31 @@
 _log = logging.getLogger(__name__)
 
 
+def _prevent_rasterization(draw):
+    # we assume that by default artists are not allowed to rasterize (unless
+    # its draw method is explicitly decorated). If it is being drawn after a
+    # rasterized artist and it has reached the rater_depth of 0. We stop
+    # rasterization so that it does not affect the behavior of normal artist
+    # (e.g., change in dpi). If the artist's draw method is decorated
+    # (draw._supports_rasterization is True), it won't  be decorated by
```
Comment (QuLogic) [inline lib/matplotlib/artist.py:30]: But the check is not for `draw._supports_rasterization is True`; it's for `hasattr(draw, '_supports_rasterization')`

File: lib/matplotlib/artist.py (hunk 3 above)
Comment (leejjoon) [reply lib/matplotlib/artist.py:33]: The only occurrence of `_supports_rasterization` is in the `Artist.set_rasterized` method (which already does ` getattr(obj, '_supports_rasterization', False)`. And I will go ahead and add `draw._supports_rasterization = False` here.

File: lib/matplotlib/artist.py (hunk 4 above)
Comment (leejjoon) [reply lib/matplotlib/artist.py:134]: Thanks. I will move this line above the comment.

File: lib/matplotlib/artist.py (hunk 5 above)
Comment (leejjoon) [reply lib/matplotlib/artist.py]: I have changed the comment. 

` If the artist's draw method is decorated (has a "_supports_rasterization" attribute),`

**Comment by leejjoon** [comment]
Somehow, `ci/circleci:docs-python38` fails, and I could not able to figure out why. Can someone take a look into this?
Once this issue is resolved, I may squash the commits into a single one.

File: lib/matplotlib/artist.py (hunk 4 above)
Comment (tacaswell) [reply lib/matplotlib/artist.py:135]: We should be more careful here about not not over-writing the `draw` method unless we have to, I will push a commit to try and fix this.

**Comment by tacaswell** [comment]
@leejjoon I took the liberty of pushing an extra commit to fix a subtle issue with (deep) class hierarchies.  I hope you do not mind.

**Comment by leejjoon** [comment]
> @leejjoon I took the liberty of pushing an extra commit to fix a subtle issue with (deep) class hierarchies. I hope you do not mind.

No, not at all.

However, the same check is also done inside the decorator,

https://github.com/matplotlib/matplotlib/blob/9de19d0b7073b79529c8c753a311090f3fe4b083/lib/matplotlib/artist.py#L33

which I think is now redundant. Do you want it to be removed?

-JJ

**Comment by ksunden** [comment]
Putting the check at this level has some subtle distinctions for subclasses and behavior of `super` calls, and in fact removes the conflicting entries from the doc builds (though the changes to address those are not worth undoing)

**Comment by leejjoon** [comment]
> Putting the check at this level has some subtle distinctions for subclasses and behavior of `super` calls, and in fact removes the conflicting entries from the doc builds (though the changes to address those are not worth undoing)

Thanks. I did not know if that will make a difference. I have removed the if statement inside the decorator.

**Review by ksunden — APPROVED** [review]

**Comment by tacaswell** [comment]
Thank you @leejjoon !

**Comment by dstansby** [comment]
👋 unfortunately this has broken some plotting in `astropy`. In `astropy.visualization.wcsaxes` we have a situation like:
```python
from matplotlib.text import Text
import matplotlib.pyplot as plt
from matplotlib.artist import Artist

class CustomArtist(Artist):
    def draw(self, renderer):
        t = CustomText(0, 0, 'a')
        t.draw(renderer, custom_param='custom_param_value')

class CustomText(Text):
    def draw(self, renderer, custom_param=None):
        super().draw(renderer)


fig, ax = plt.subplots()
ax.add_artist(CustomArtist())
plt.show()
```

The key here is the dummy `custom_param` (in astropy this parameter is used). Before this PR everything was fine (my sample code above fails, but on a different error), but after this PR the error below is raised. I think `astropy` is doing Bad Things by having parameters in the overriden `draw()` methods, and I am working on changing it (https://github.com/astropy/astropy/issues/12672) but making slow progress.

Would be very greatful if there's an easy way to fix this for `astropy`. If not, and we need to fix on our end by removing extra arguments to `draw()`, please let me know and I can use that as motivation to get that change done quicker!

```python
Traceback (most recent call last):
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/backend_bases.py", line 1212, in _on_timer
    ret = func(*args, **kwargs)
          ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/backends/backend_macosx.py", line 68, in callback_func
    callback()
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/backends/backend_macosx.py", line 88, in _draw_idle
    self.draw()
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/backends/backend_macosx.py", line 50, in draw
    super().draw()
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/backends/backend_agg.py", line 400, in draw
    self.figure.draw(self.renderer)
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/artist.py", line 95, in draw_wrapper
    result = draw(artist, renderer, *args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/artist.py", line 72, in draw_wrapper
    return draw(artist, renderer)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/figure.py", line 3077, in draw
    mimage._draw_list_compositing_images(
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/image.py", line 131, in _draw_list_compositing_images
    a.draw(renderer)
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/artist.py", line 72, in draw_wrapper
    return draw(artist, renderer)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/axes/_base.py", line 3148, in draw
    mimage._draw_list_compositing_images(
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/image.py", line 131, in _draw_list_compositing_images
    a.draw(renderer)
  File "/Users/dstansby/github/matplotlib/lib/matplotlib/artist.py", line 39, in draw_wrapper
    return draw(artist, renderer)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dstansby/github/matplotlib/test.py", line 8, in draw
    t.draw(renderer, custom_param='custom_param_value')
TypeError: CustomText.draw() got an unexpected keyword argument 'custom_param'
```

**Comment by tacaswell** [comment]
The wrapper should blindly forward `*args, **kwargs`, sorry for not thinking of this case on review.

**Comment by tacaswell** [comment]
```diff
diff --git a/lib/matplotlib/artist.py b/lib/matplotlib/artist.py
index a4693f78cf..7967f597d4 100644
--- a/lib/matplotlib/artist.py
+++ b/lib/matplotlib/artist.py
@@ -29,14 +29,14 @@ def _prevent_rasterization(draw):
     # (e.g., change in dpi).
 
     @wraps(draw)
-    def draw_wrapper(artist, renderer):
+    def draw_wrapper(artist, renderer, *args, **kwargs):
         if renderer._raster_depth == 0 and renderer._rasterizing:
             # Only stop when we are not in a rasterized parent
             # and something has been rasterized since last stop.
             renderer.stop_rasterizing()
             renderer._rasterizing = False
 
-        return draw(artist, renderer)
+        return draw(artist, renderer, *args, **kwargs)
 
     draw_wrapper._supports_rasterization = False
     return draw_wrapper

```

should be enough, working on a PR.

**Comment by tacaswell** [comment]
```python
from matplotlib.text import Text
import matplotlib.pyplot as plt
from matplotlib.artist import Artist

class CustomArtist(Artist):
    def draw(self, renderer):
        t = CustomText(0, 0, 'a', figure=self.figure)
        t.draw(renderer, custom_param='custom_param_value')

class CustomText(Text):
    def draw(self, renderer, custom_param=None):
        super().draw(renderer)


fig, ax = plt.subplots()
ax.add_artist(CustomArtist())
plt.show()

```

Fixes the other issue.

**Pull request merged**